uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/federatedlogin/testscripts/FederatedLogin.java

```diff
@@ -0,0 +1,350 @@
+                    new byte[]{127, 0, 0, 1})
+                    && Arrays.equals(Inet4Address.getByName("testzone3.localhost").getAddress(),
+                    new byte[]{127, 0, 0, 1});
+        } catch (UnknownHostException e) {
+            return false;
+        }
+    }
+
+    @Test
+    public void testCrossZoneSamlIntegration() throws Throwable {
+        String idpZoneId = "testzone1";
+        String idpZoneUrl = baseUrl.replace("localhost", idpZoneId + ".localhost");
+        String spZoneId = "testzone2";
+        String spZoneUrl = baseUrl.replace("localhost", spZoneId + ".localhost");
+        RestTemplate adminClient = getAdminClient();
+        RestTemplate identityClient = getIdentityClient();
+        //Creating Orch IDP Zone
+        IdentityZone idpZone = IntegrationUtilsOrchestrator.createOrchZone(identityClient, baseUrl, idpZoneId, idpZoneId, null);
+        idpZone.getId();
+        String idpZoneUserEmail = new RandomValueStringGenerator().generate() + "@samltesting.org";
+        //Create user for IDP Admin
+        createZoneUser(idpZoneId, idpZoneUserEmail, idpZoneUrl);
+        SamlConfig samlConfig = new SamlConfig();
+        samlConfig.setWantAssertionSigned(true);
+        samlConfig.addAndActivateKey("key-1", new SamlKey(key1, passphrase1, certificate1));
+        samlConfig.addKey("key-2", new SamlKey(key2, passphrase2, certificate2));
+        IdentityZoneConfiguration config = new IdentityZoneConfiguration();
+        config.setSamlConfig(samlConfig);
+        //Creating Orch SP Zone
+        IdentityZone spZone = IntegrationUtilsOrchestrator.createOrchZone(identityClient, baseUrl, spZoneId, spZoneId, config);
+        //Get Client credentials to for SP Admin token
+        String spZoneAdminToken = IntegrationTestUtils.getClientCredentialsToken(serverRunning, "admin", "adminsecret");
+        //Get IDP meta Data
+        SamlIdentityProviderDefinition samlIdentityProviderDefinition = createZone1IdpDefinition(IDP_ENTITY_ID);
+        //configure IDP with Metadata
+        IdentityProvider<SamlIdentityProviderDefinition> idp = getSamlIdentityProvider(spZone.getId(), spZoneAdminToken, samlIdentityProviderDefinition);
+        //Get Sp meta data
+        SamlServiceProviderDefinition samlServiceProviderDefinition = createZone2SamlSpDefinition("cloudfoundry-saml-login");
+        //Configure SP with Metadata
+        SamlServiceProvider service = getSamlServiceProvider(idpZone.getId(), spZoneAdminToken, samlServiceProviderDefinition, "testzone2.cloudfoundry-saml-login", "Local SAML SP for testzone2", baseUrl);
+        //Login into SP with IDP credentials
+        performLogin(idpZone.getId(), idpZoneUserEmail, idpZoneUrl, spZone, spZoneUrl, samlIdentityProviderDefinition);
+        webDriver.get(baseUrl + "/logout.do");
+        webDriver.get(spZoneUrl + "/logout.do");
+    }
+
+    private RestTemplate getAdminClient() {
+        return IntegrationUtilsOrchestrator.getClientCredentialsTemplate(
```

Comment:
Duplicate again.  You might want to review the difference between IntegrationUtilsOrchestrator and IntegrationTestUtils.  Use functions from IntegrationTestUtils whenever possible